src/main/java/com/couchbase/lite/router/BufferOutputStream.java

```diff
@@ -0,0 +1,31 @@
+package com.couchbase.lite.router;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.LinkedList;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class BufferOutputStream extends OutputStream {
+
+    LinkedList<Byte> buffer = new LinkedList<Byte>();
```

Comment:
I don't think that it much of a perfromance issue since it is only appending at the end and removing from the first position. But you are right, an array would be more memory efficient.